### PR TITLE
[#402] Support Floating x or x.y version is Artemis CRD version string

### LINKED ISSUE #402: Support Floating x or x.y version is Artemis CRD version string
**Is your feature request related to a problem? Please describe.**
The Version of the Artemis CRD allows a broker version to be specified, in the absence of this being set, the operator picks the latest or current version. When it is present it locks down the version.

However it would be better if the version could also specify a floating version, like 7 or 7.2 or 7.2.1 and in all cases if the match is within a range (as configured in the operator) the latest in that range is used.

**Describe the solution you'd like**

**Describe alternatives you've considered**
Optional. A clear and concise description of any alternative solutions or features you've considered.

**Additional context**
Optional. Add any other context or screenshots about the feature request here.

### PR DESCRIPTION
**Pull request by gaohoward**

### CODE REVIEW

File: controllers/activemqartemis_reconciler.go [hunk 1]
```diff
@@ -2200,10 +2214,14 @@ func configPodSecurity(podSpec *corev1.PodSpec, podSecurity *brokerv1beta1.PodSe
 	}
 }
 
-func determineImageToUse(customResource *brokerv1beta1.ActiveMQArtemis, imageTypeName string) string {
+func determineImageToUse(customResource *brokerv1beta1.ActiveMQArtemis, imageTypeName string) (string, error) {
 
 	imageName := ""
-	compactVersionToUse := determineCompactVersionToUse(customResource)
+	compactVersionToUse, err := determineCompactVersionToUse(customResource)
```
Comment (gtully) [inline controllers/activemqartemis_reconciler.go:2220]: I am not seeing the value of error handling here.... seems empty string is the default in the case of error or unknown.
if we drop the error no use computing it.

File: controllers/activemqartemis_controller_test.go [hunk 2]
```diff
@@ -118,6 +119,203 @@ var _ = Describe("artemis controller", func() {
...
+			resolved = common.ResolveBrokerVersion(versionsMap, requested)
+			Expect(resolved).To(BeNil())
+
+			requested, _ = semver.NewVersion("8")
+			resolved = common.ResolveBrokerVersion(versionsMap, requested)
+			Expect(resolved).To(BeNil())
+
+		})
+
+		It("version ignored if images are explicitly specified", func() {
```
[the remark this answers is not in the document]
Comment (gaohoward) [reply controllers/activemqartemis_controller_test.go:183]: yeah good point, will add a validation.

File: controllers/activemqartemis_controller_test.go [hunk 3]
```diff
@@ -118,6 +119,203 @@ var _ = Describe("artemis controller", func() {
...
+			defer func() {
+				os.Unsetenv("RELATED_IMAGE_ActiveMQ_Artemis_Broker_Kubernetes_" + version.CompactLatestVersion)
+				os.Unsetenv("RELATED_IMAGE_ActiveMQ_Artemis_Broker_Init_" + version.CompactLatestVersion)
+			}()
+			By("deploy a broker")
+			verFields := strings.Split(version.LatestVersion, ".")
+			major := verFields[0]
+			brokerCr, createdBrokerCr := DeployCustomBroker(defaultNamespace, func(candidate *brokerv1beta1.ActiveMQArtemis) {
+				candidate.Spec.Version = major
+				candidate.Spec.DeploymentPlan.Image = ""
```
[the remark this answers is not in the document]
Comment (gaohoward) [reply controllers/activemqartemis_controller_test.go:223]: +1

File: controllers/activemqartemis_reconciler.go (hunk 1 above)
Comment (gaohoward) [reply controllers/activemqartemis_reconciler.go:2220]: yep, make sense.

**Comment by gtully** [comment]
great.

I wonder, if we are to do some version aware code in the operator, do we need to drop the image url option? Or do we have a way to extract the version from the image?

If we can't easily infer the version from the image, and if it is a custom image I think we can't without some authenticated query to the registry, do we need to support some sort of "-final" postfix to a version string to say - no updates.
Even that will be problematic if the mapped image url changes under the hood, we will need to annotate the CR to mark the first image we resolved to in the -final case. This may not be ideal.

some food for thought.

**Review by gtully — APPROVED** [review]

**Comment by gaohoward** [comment]
> great.
> 
> I wonder, if we are to do some version aware code in the operator, do we need to drop the image url option? Or do we have a way to extract the version from the image?
> 
> If we can't easily infer the version from the image, and if it is a custom image I think we can't without some authenticated query to the registry, do we need to support some sort of "-final" postfix to a version string to say - no updates. Even that will be problematic if the mapped image url changes under the hood, we will need to annotate the CR to mark the first image we resolved to in the -final case. This may not be ideal.
> 
> some food for thought.

Yeah there are certainly places to improve. I put the though into this https://github.com/artemiscloud/activemq-artemis-operator/issues/430 for tracking and discussion.

Thanks!

**Pull request merged**